### PR TITLE
Text Assessment V2

### PR DESCRIPTION
**Pull request by jpbernius**
### Checklist

- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Current Text Assessment is problematic in many ways:

1. mapping between Text and Feedback is cumbersome and
2. implementation is inconsistent with e.g. assessment of modeling exercises.

### Description

This PR introduces a new UI to add feedback to textblocks inline.
Tutors can hover of the precomputed Text Blocks and click to add a Feedback Element inline.
When editing a feedback element, the referencees text block is highlighted.

### Steps for Testing

1. Log in to Artemis
2. Create Text Submission, make sure to use sentences with punctuation so it runs without segmentation).
3. Open a Text Assessment
4. Navigate to New Assement UI by Changing URL from `/#/course-management/:exerciseId/text-exercises/:exerciseId/submissions/:submissionId/assessment`
   to `/#/course-management/:exerciseId/text-exercises/:exerciseId/submissions`**`-new`**`/:submissionId/assessment`
5. Hover over text to selet block, click to edit feedback and save/submit.

### Screenshots

ta-v2

### Future Work

- Example Submissions
- Complaints
- Custom definitions of Text Blocks

### CODE REVIEW

**Comment by jpbernius** [comment]
Failing build is related to nodejs/nodejs.org#3075 (Their infra is down, so the @github @actions cannot install Node.

**Comment by jpbernius** [comment]
@sascha11110: Which Browser are you using. I know that the surrounding box gets enlarged in some browsers.

**Comment by sascha11110** [comment]
I've tested with Chrome on macOS. 

But also in Safari there is a minor problem when one inserts a feedback element at the bottom:

image

**Comment by clemenszuck** [comment]
- I also got the visual bug mentioned by @sascha11110 using Windows and Chrome, Firefox or Microsoft Edge.

- In Microsoft Edge the Score field does not have the arrows to change the score using the mouse, so you have to use the keyboard to enter the score: 
assessment-v2

- One minor issue regarding usability I noticed is that you can not remove feedback boxes which have a feedback text or a score other than 0. So if you want to remove an older (green) feedback box completely you have to delete the text and set the score to 0. In my opinion it would make it a little easier to use if the "X" to remove a box would be always visible.

**Comment by jpbernius** [comment]
Thanks for your feedback and tests @clemenszuck and @sascha11110.

---

>I also got the visual bug mentioned by @sascha11110 using Windows and Chrome, Firefox or Microsoft Edge.

This is fixed with 2e50b60. Thanks for noticing. :)

---

> In Microsoft Edge the Score field does not have the arrows to change the score using the mouse, so you have to use the keyboard to enter the score:

Would it help to follow what @sascha11110 suggested:

> Maybe you could make the two input fields for "Feedback" and "Score" a little bit more visible. Currently they are not distinguishable from the surrounding box.

I don't think we need to implement the buttons ourselves, or what does everyone think on that?

---

> One minor issue regarding usability I noticed is that you can not remove feedback boxes which have a feedback text or a score other than 0.

So this was a design decision as one should not delete feedback by accident. What about this idea:
We always show the X. If the feedback is not empty, we put up a `delete` label which one needs to click again.
This would make it easier but still put some extra step before discarding feedback.

**Comment by krusche** [comment]
> So this was a design decision as one should not delete feedback by accident. What about this idea:
> We always show the X. If the feedback is not empty, we put up a `delete` label which one needs to click again.
> This would make it easier but still put some extra step before discarding feedback.

You could reuse the way student questions and student answers can be deleted. This is also a two step process to prevent accidentally deleting something

**Comment by clemenszuck** [comment]
> > Maybe you could make the two input fields for "Feedback" and "Score" a little bit more visible. Currently they are not distinguishable from the surrounding box.

> I don't think we need to implement the buttons ourselves, or what does everyone think on that?

Making the input fields more visible would definitely help with that. Implementing these buttons is certainly not necessary, it's just a little strange that they are present in the other browsers but not in Microsoft Edge. Is it a known limitation of Edge?

> So this was a design decision as one should not delete feedback by accident. What about this idea:
We always show the X. If the feedback is not empty, we put up a delete label which one needs to click again.
This would make it easier but still put some extra step before discarding feedback.

That makes sense. The idea sounds perfect!

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java [hunk 1]
```diff
@@ -190,11 +209,13 @@ public TextAssessmentResource(AuthorizationCheckService authCheckService, Result
      * Splits the TextSubmission corresponding to a resultId into TextBlocks.
      * The TextBlocks get a suggested feedback if automatic assessment is enabled and feedback available
      *
+     * @deprecated Text Blocks should be returned with Submission.
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:212]:
```suggestion
     * @deprecated Text Blocks should be returned with Submission. Please use V2 text assessment which does not need this API call. This API call will be removed soon.
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java [hunk 2]
```diff
@@ -285,17 +307,22 @@ private void computeBlocks(Result result, TextExercise textExercise) {
         final List<Feedback> assessments = textAssessmentService.getAssessmentsForResult(result);
         result.setFeedbacks(assessments);
 
-        if (exercise.isAutomaticAssessmentEnabled() && (textSubmission.getBlocks() == null || textSubmission.getBlocks().isEmpty())) {
-            computeBlocks(result, exercise);
+        final List<TextBlock> textBlocks = textBlockRepository.findAllBySubmissionId(textSubmission.getId());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:310]: add some documentation comments

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/TextAssessmentResource.java:310]: Also explain the fall back solution here

**Pull request merged**